### PR TITLE
Fixed a bug in variation event detection code that could sometimes lead to mistreating indel assembly windows as SNP assembly windows

### PR DESCRIPTION
**Pull request by jamesemery**
Noticed this while drilling into sites for #6634. Pulling it out since this is a separate bugfix. On occasion this would result in funky hmm trimming behavior (like only trimming to 20 bases on the edges of an indel when we should have trimmed to 75) and possibly other bizarre issues.

### CODE REVIEW

**Comment by jamesemery** [comment]
Also I'm slightly afraid of this branch on the grounds that it makes subtle math rounding changes here and there in the expected outputs.

**Comment by davidbenjamin** [comment]
@jamesemery I see how the returned `SortedSet` in the existing code is not sorted with respect to the resulting trimming bounds, but I don't see how it leads to a bug.  Here's the code in `AssemblyRegionTrimmer::trim` that the `SortedSet` of all variation events gets passed to:

```
for (final VariantContext vc : variantsInRegion) {
   int padding = assemblyRegionArgs.snpPaddingForGenotyping;
   if (vc.isIndel()) {
      padding = assemblyRegionArgs.indelPaddingForGenotyping;
      // omitting a few lines where STRs get additional padding
   }

   // seems to me like if this is the second event and it's an indel then minStart could extend
   // past the padding of a previous SNP event
   minStart = Math.min(minStart, Math.max(vc.getStart() - padding,1));
   maxEnd = Math.max(maxEnd, vc.getEnd() + padding);
}

final SimpleInterval paddedVariantSpan = new SimpleInterval(region.getContig(), minStart, maxEnd).intersect(region.getPaddedSpan());
```

**Comment by jamesemery** [comment]
@davidbenjamin The bug here is that we misinterpreted what a TreeSet in java does. The actual behavior for this method was that it only took the FRIST variant at each start position that it saw from the ordering of the haplotypes it saw. This meant if a SNP and INDEL started at the same position then there was a chance that site only looks like a SNP and we trim incorrectly. Look at this from the TreeSet object docs:
```<p>Note that the ordering maintained by a set (whether or not an explicit
 * comparator is provided) must be <i>consistent with equals</i> if it is to
 * correctly implement the {@code Set} interface.  (See {@code Comparable}
 * or {@code Comparator} for a precise definition of <i>consistent with
 * equals</i>.)  This is so because the {@code Set} interface is defined in
 * terms of the {@code equals} operation, but a {@code TreeSet} instance
 * performs all element comparisons using its {@code compareTo} (or
 * {@code compare}) method, so two elements that are deemed equal by this method
 * are, from the standpoint of the set, equal.  The behavior of a set
 * <i>is</i> well-defined even if its ordering is inconsistent with equals; it
 * just fails to obey the general contract of the {@code Set} interface.```

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyResultSet.java [hunk 1]
```diff
@@ -519,13 +520,32 @@ public void regenerateVariationEvents(int maxMnpDistance) {
...
-        final TreeSet<VariantContext> vcs = new TreeSet<>(Comparator.comparingInt(VariantContext::getStart));
+        final TreeSet<VariantContext> vcs = new TreeSet<>(
+                Comparator.comparingInt(VariantContext::getStart)
+                        // Decide arbitrarily so as not to accidentally throw away overlapping variants
+                .thenComparing(new Comparator<VariantContext>() {
+                    @Override
+                    public int compare(VariantContext v1, VariantContext v2) {
+                        List<Allele> v1Alleles = v1.getAlleles();
+                        List<Allele> v2Alleles = v2.getAlleles();
+                        Utils.validate(v1Alleles.size() == 2, () -> "Error Haplotype event map Variant Context has too many alleles: "+v1);
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyResultSet.java:537]: This feels both too `private` for the validation to be necessary and too deep for it to be helpful.  Could you find a higher entry point to check that the variant contexts in the `EventMap` are all biallelic, if one does not exist already?

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyResultSet.java [hunk 2]
```diff
@@ -519,13 +520,32 @@ public void regenerateVariationEvents(int maxMnpDistance) {
...
+                        Utils.validate(v2Alleles.size() == 2, () -> "Error Haplotype event map Variant Context has too many alleles: "+v2);
+
+                        int diff = v1.getReference().length() - v2.getReference().length();
+                        if (diff == 0) {
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyResultSet.java:541]: It seems like the `if (diff == 0)` block would be more clearly expressed as a second `thenComparing`:
```
Comparator.comparingInt(VariantContext::getStart)
   .thenComparing(Comparator.comparingInt(vc -> vc.getReference().length()))
   .thenComparing(alt-allele-based tiebreaker)
```

**Comment by davidbenjamin** [comment]
Thanks for the explanation!

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyResultSet.java (hunk 2 above)
Comment (jamesemery) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyResultSet.java:541]: you are right, done.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyResultSet.java (hunk 1 above)
Comment (jamesemery) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyResultSet.java:537]: I pushed this up into event map construction.

**Comment by jamesemery** [comment]
@davidbenjamin Back to you. Now the code is very small indeed. Perhaps its worth writing a test of this code however....

**Comment by jamesemery** [comment]
@ldgauthier Can you take a look at this? This fixed a pretty rare (but not rare enough that it wasn't causing mismatches in my other work) error in GATK where we would be mis-sizing assembly windows where indels that overlapped snps were involved. It made some moderate differences in PL scores in HC tests which suggests its actually happening somewhat frequently. I see however that it affected one of the CombineGVCFs integration tests (specifically all the values seem to be off by very small margins). Is that a problem at all?

**Comment by ldgauthier** [comment]
The CombineGVCF changes are because I use the HC expected output as an input.  (Pros and cons.)  All the changes in the NA12878 GVCF are mirrored in the CombineGVCFs changes, so I think that's all working as expected.

File: src/test/resources/org/broadinstitute/hellbender/tools/walkers/CombineGVCFs/testBreakBandsArgumet.vcf [hunk 3]
```diff
@@ -8,7 +8,7 @@
...
...t  --convert-to-base-pair-resolution false --input-is-somatic false --drop-somatic-filtering-annotations false --ignore-variants-starting-outside-interval false --combine-variants-distance 0 --max-distance 2147483647 --ref-padding 1 --interval-set-rule UNION --interval-padding 0 --interval-exclusion-padding 0 --interval-merging-rule ALL --read-validation-stringency SILENT --seconds-between-progress-updates 10.0 --disable-sequence-dictionary-validation false --create-output-bam-index true --create-output-bam-md5 false --create-output-variant-index true --create-output-variant-md5 false --lenient false --add-output-sam-program-record true --add-output-vcf-command-line true --cloud-prefetch-buffer 40 --cloud-index-prefetch-buffer -1 --disable-bam-index-caching false --sites-only-vcf-output false --help false --version false --showHidden false --QUIET false --use-jdk-deflater false --use-jdk-inflater false --gcs-max-retries 20 --gcs-project-for-requester-pays  --disable-tool-default-read-filters false --disable-tool-default-annotations false --enable-all-annotations false --allow-old-rms-mapping-quality-annotation-data false",Version="Unavailable",Date="August 10, 2020 4:07:09 PM EDT">
```
[the remark this answers is not in the document]
Comment (jamesemery) [reply src/test/resources/org/broadinstitute/hellbender/tools/walkers/CombineGVCFs/testBreakBandsArgumet.vcf:11]: I assumed that you left the headers in for documentation purposes. If that is not the case then i'll get rid of them...

**Comment by jamesemery** [comment]
@davidbenjamin can you take another look at this? I would like to get it in before #6634 goes in

**Comment by davidbenjamin** [comment]
@jamesemery :thumbsup:

**Pull request merged**